### DISCUSSION TITLE
Strange results with HGST vs seagate drives

### DISCUSSION THREAD

**Discussion by Jazz9** [minio/minio#16832] [Q&A] [answered]
## Expected Behavior
Consistent behavior across drives 

## Current Behavior
Currently, the HGST drives don't do flushes and perform very well. The Seagate EXOS and Ironwolf Pro drives get flushed by the OS and perform badly. We see this in the same server / controller and also in completely different hardware.

![image](https://user-images.githubusercontent.com/39378577/225578901-7283c783-2957-4515-a12c-5f8117beec17.png)

## Steps to Reproduce (for bugs)
Put the drives in the server and format with XFS. Then attempt to write to the drives with minio.

We suspect that minio is sometimes using o_direct and other times sync but wanted to know if you can configure it?

## Your Environment
ubuntu 22
xfs formatted drives
18TB Exos drives or 8TB HGST drives.
supermicro with LSI HBA and also with HP dl380g9 p440ar

**Comment by Jazz9** [comment]
After more investigation, it seems the hgst drives are 4kn and the seagate are 512e. Both formatted with the default xfs settings and both support o_direct.

**Comment by klauspost** [comment, marked as answer]
MinIO has no idea what hardware lies below the filesystem. We always flush small writes with fdatasync and use o_direct for bigger writes.

I can't tell you why this difference shows up.

**Reply by Jazz9** [reply]
thankyou for the response. Can you point me to the code that controls this? I'd like to review it and see if I can find out why it is incorrectly choosing to use fdatasync.

**Reply by harshavardhana** [reply]
It will never incorrectly choose fdatasync you just need to throw away your Seagate drives that's all.

**Reply by Jazz9** [reply]
It's not a seagate issue - its 4kn vs 512e drives.

**Reply by harshavardhana** [reply]
We really don't care @Jazz9 not our problem, simply choose what works best for your requirements.

If you need deeper architecture reviews sign up for subscription https://min.io/pricing

**Reply by Jazz9** [reply]
No worries - I have already engaged the minio sales team and are currently trialing the solution.

**Reply by Jazz9** [reply]
Also, it is incorrectly using fdatasync but it seems likely a xfs issue ( there is a potential patch that may fix this ). We have 300+ drives that would be affected so i'm trying to not replace them :D

**Reply by harshavardhana** [reply]
@Jazz9 here is what we do always without fail during PUT calls

- any object that is smaller than 128KiB is written directly to disk in a single call. The file is opened via O_SYNC so there is no fdatasync the I/O is fully synchronous here

- any object greater than this goes via a different approach
   - xl.meta (O_SYNC) with part.1 (O_DIRECT for all aligned writes and any remaining one gets flushed via fdatasync to disk)
   
This is true for all I/O in MinIO

**Reply by Jazz9** [reply]
interesting - I will continue to investigate the issue and let you know if I come to a conclusion.

**Comment by Jazz9** [comment]
Just a followup for anyone that lands here from google - 

The issue was not with Minio but with the ubuntu disk scheduler. With some tuning, we were able to improve the write speed ~ 30-50% . Minio was doing all the correct things.

**Reply by harshavardhana** [reply]
What is the change in scheduler? And what was default? And what was the kernel version?

**Reply by Jazz9** [reply]
Linux nas-aubun-rk6-min01 5.15.0-58-generic
XFS drives
testing against 3 different hardware setups / drives / controllers

hdparm to disable write cache on the drives ( this didn't make a large difference but for the safety of file we feel this is important. )

echo "write through" | sudo tee /sys/block/sd*/queue/write_cache
default was write back
We have a write heavy workload and this more than doubled the throughput at the client S3 layer.

The disk scheduler was changed to none but the difference there wasn't enough to warrant the change from non-standard setup.

I have another change that we believe would speed up performance but would require a minio code change. If I do up a proposal, would there be someone that would interested in reading?
